GlyGen DCC API Enpoints

HadleyKing opened the pull request:
Enabled 1st API endpoint for [GlyGen](https://api.glygen.org/#/Protein_Search_(generic_protein_API)/get_directsearch_protein__query__query_json_). This incorporates the two queries listed by GlyGen team on the [DCC API Endpoints sheet](https://docs.google.com/spreadsheets/d/1oNCfD90R3-QNS3Si5nQxfOolpM---k8eAbxb-le9qvI/edit?urp=gmail_link#gid=0) added by @jeet-vora and @kee007ney

HadleyKing committed:
GlyGen DCC API  Endpoints

Fix #1. Enabled 1st API endpoint for GlyGen.
On branch gene_to_glycosylation
Your branch is up to date with 'origin/gene_to_glycosylation'.
Changes to be committed:
	new file:   components/gly_gen/index.tsx
	new file:   components/gly_gen/package.json
	modified:   components/index.ts
	modified:   package-lock.json

HadleyKing committed:
GlyGen DCC API  Endpoints

Fix #1. Enabled 1st API endpoint for GlyGen.
On branch gene_to_glycosylation
Your branch is up to date with 'origin/gene_to_glycosylation'.
Changes to be committed:
	new file:   components/gly_gen/index.tsx
	new file:   components/gly_gen/package.json
	modified:   components/index.ts
	modified:   package-lock.json

HadleyKing committed:
Merge pull request #2 from biocompute-objects/gene_to_glycosylation

GlyGen DCC API  Endpoints

HadleyKing committed:
Merge branch 'nih-cfde-main'

HadleyKing committed:
Merge branch 'main' of github.com:biocompute-objects/playbook-partnership

u8sand commented on components/gly_gen/index.tsx:
```diff
@@ -0,0 +1,51 @@
+  .meta({
+    label: 'GlyGen Query',
+    description: 'Search for records in GlyGen. Must be a JSON input',
+  })
```
Since this object needs to be created before you can compute the protein product information, it's important that there is a way for an end user to create it, i.e. through a Prompt process.

It might be helpful if at the very least you include an `example` in `meta` with an example GlyGenQuery object.

u8sand commented on components/gly_gen/index.tsx:
```diff
@@ -0,0 +1,51 @@
+
+export type GlyGenResponseType = {
+  mydatatype: string[]
+}
```
The GlyGenDataType & GlyGenResponseTypes do not seem correct? These should define the shape of the data your API expects, your API docs suggest the query looks something like `{"uniprot_canonical_ac":"P14210"}`, the type of this would be:
```ts
type GlyGenDataType = {
  uniprot_canonical_ac: string
}
```

The response looked very different as well. One thing we started doing in other components is *runtime type checking* as well. https://github.com/nih-cfde/playbook-partnership/blob/main/components/gene_count_matrix/index.tsx#L13-L20 for example. this will show an error in the prototype interface when the type is wrong.

The request is more important than the response at this point so you don't have to make it "complete" -- it could just be:
```
t.type({ queryinfo: t.any, results: t.array(t.any) })
```
for example, but it should not be conflicting with the true type of the actual input/output.

HadleyKing replied:
@u8sand Thanks so much for the feedback!! You have answered all the questions I had about this process but did not have a chance to ask yet.

u8sand replied:
Happy to help any way I can to drive this further. Thank you for your contribution!

u8sand reviewed (changes requested):
A great start! Unfortunately there are a few things we'll want to address to better integrate this component.

Most notably, the component is "orphaned" and there is no way for a user to actually discover/access it. Either there should be a [Prompt](https://github.com/nih-cfde/playbook-partnership/blob/main/docs/contributions.md#prompt) to facilitate user construction of the GlyGenQuery, or there should be a way to derive a GlyGenQuery from some existing input (currently, File or GeneSymbol)

HadleyKing committed:
saving [WIP]
Changes to be committed:
	modified:   components/gly_gen/index.tsx
	modified:   components/index.ts
	modified:   package-lock.json

HadleyKing committed:
saving [WIP]
Changes to be committed:
	modified:   components/gly_gen/index.tsx
	modified:   components/index.ts
	modified:   package-lock.json

HadleyKing committed:
Merge branch 'nih-cfde-main'

HadleyKing committed:
Merge branch 'main' of github.com:biocompute-objects/playbook-partnership

HadleyKing committed:
update packages

HadleyKing committed:
Add GlyGenResponseNode that takes GeneInfo

Changes to be committed:
	modified:   components/gly_gen/index.tsx
	modified:   components/index.ts
	modified:   package-lock.json

u8sand commented on components/gly_gen/index.tsx:
```diff
@@ -0,0 +1,49 @@
+        accept: 'application/json'
+      }
+      })
+    return response.json()
```
So the way `fetch` works in javascript is in two steps I usually write as follows:

```tsx
const request = await fetch('...')
const response = await request.json()
// your data is now in response
```

Namely you've returned `response.json()` which is a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise), but you need to `await` it to get the actual data.

HadleyKing replied:
I modify those lines to be:

```
  .resolve(async (props) => {
    const query = encodeURIComponent(`{"recommended_gene_name":"${props.inputs.gene.symbol}"`)
    console.log('props', props, query)
    const request = await fetch(`https://api.glygen.org/directsearch/protein/?query=${query}`, {
      method: 'GET',
      headers: {
        accept: 'application/json'
      }
      })
    const response = await request.json()
    console.log('response', response)
```

And I get in the terminal (from console.log):
```
props {
  inputs: {
    gene: {
      _id: '59272',
      symbol: 'ACE2',
      entrezgene: '59272',
      name: 'angiotensin converting enzyme 2',
      taxid: 9606,
      ensembl: [Object]
    }
  }
} %7B%22recommended_gene_name%22%3A%22ACE2%22
response { error_list: [ { error_code: 'invalid-query-json' } ] }
props {
  inputs: {
    gene: {
      _id: '59272',
      symbol: 'ACE2',
      entrezgene: '59272',
      name: 'angiotensin converting enzyme 2',
      taxid: 9606,
      ensembl: [Object]
    }
  }
} %7B%22recommended_gene_name%22%3A%22ACE2%22
response { error_list: [ { error_code: 'invalid-query-json' } ] }
```

I seem to have something mixed up with the type script object declarations? I am new to type script but not so much react. 

The application does not look like it ever pings the glygen site.... 

![Screen Shot 2022-12-14 at 6 40 23 PM](https://user-images.githubusercontent.com/20343214/207738780-d511f945-4675-479b-abf1-54cdc9ef0aa8.png)

u8sand replied:
So the actual API request happens server side so you won't see it in the console unfortunately.

In the code you provided there seems to be an error:

```ts
    const query = encodeURIComponent(`{"recommended_gene_name":"${props.inputs.gene.symbol}"`)
```

Namely you're missing another `}` here:
```
ol}"`)
```
 =>
```
ol}"}`)
```

I recommend:
```ts
const query = encodeURIComponent(JSON.stringify({
  recommended_gene_name: props.inputs.gene.symbol,
}))
```
And let javascript make sure the json is valid

HadleyKing committed:
GlyGen DCC API Enpoints Review #2
Changes to be committed:
	modified:   components/gly_gen/index.tsx

HadleyKing committed:
Update based on review

u8sand commented on components/gly_gen/index.tsx:
```diff
@@ -0,0 +1,50 @@
+  .codec(GlyGenResponse)
+  .view(data => (
+    <>
+      <div><pre>{JSON.parse(data.results)}</pre></div>
```
data.results is already parsed by the time it gets here, trying to re-parse it during rendering is the reason this crashes in production. it should actually be:
```tsx
<pre>{JSON.stringify(data.results)}</pre>
```
Going from data to string

HadleyKing committed:
https://github.com/nih-cfde/playbook-partnership/pull/8#pullrequestreview-1219580353

Changes to be committed:
	modified:   components/gly_gen/index.tsx

HadleyKing committed:
Added stringify

u8sand commented on components/gly_gen/index.tsx:
```diff
@@ -0,0 +1,50 @@
+      }
+      })
+    const response = await request.json()
+  })
```
It looks like you forgot to return the response in the `resolve` function, that causes an error elsewhere which expects json to come out of this method.

```tsx
return response
```

HadleyKing replied:
That is embarrassing. Thanks for your patience!

HadleyKing committed:
Final commit (hopefully)
Changes to be committed:
	modified:   components/gly_gen/index.tsx

u8sand reviewed (changes requested):
Please incorporate the provided patch, after this I will merge it:

```diff
diff --git a/components/gly_gen/index.tsx b/components/gly_gen/index.tsx
index aec9ae4..d879409 100644
--- a/components/gly_gen/index.tsx
+++ b/components/gly_gen/index.tsx
@@ -3,13 +3,17 @@ import { MetaNode } from '@/spec/metanode'
 import { GeneInfo } from '../service/mygeneinfo'
 import { z } from 'zod'
 
-export const GlyGenGeneType = {
-  gene_name: typeof GeneInfo
-}
-
 export const GlyGenResponse = z.object({
-  queryinfo: z.any(),
-  results: z.array(z.any())
+  queryinfo: z.object({
+    query: z.object({
+      recommended_gene_name: z.string()
+    }),
+  }),
+  results: z.array(
+    z.object({
+      uniprot_canonical_ac: z.string(),
+    })
+  ),
 })
 
 export type GlyGenResponseType = z.infer<typeof GlyGenResponse>
@@ -53,7 +57,7 @@ export const ProteinProductInformation = MetaNode.createProcess('ProteinProductI
         accept: 'application/json'
       }
       })
-    const response = await request.json()
+    const response = GlyGenResponse.parse(await request.json())
     return response
   })
   .build()
\ No newline at end of file
```

In general it's best to avoid `z.any()`, instead we explicitly define the shape our our type and assert it when we receive information from the API.

Thus if the API returns something unexpected, this will fail at the `resolve` step rather than at the `render` step and the render function is now strictly type safe.

HadleyKing committed:
requested changes
Changes to be committed:
	modified:   components/gly_gen/index.tsx
